x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java

```diff
@@ -657,6 +661,91 @@ public void getUserPrivileges(AuthorizationInfo authorizationInfo, ActionListene
         }
     }
 
+    @Override
+    public void getRemoteAccessRoleDescriptorsIntersection(
+        final String remoteClusterAlias,
+        final AuthorizationInfo authorizationInfo,
+        final ActionListener<RoleDescriptorsIntersection> listener
+    ) {
+        if (authorizationInfo instanceof RBACAuthorizationInfo == false) {
+            listener.onFailure(
+                new IllegalArgumentException("unsupported authorization info:" + authorizationInfo.getClass().getSimpleName())
+            );
+            return;
+        }
+
+        final Role role = ((RBACAuthorizationInfo) authorizationInfo).getRole();
+        final RemoteIndicesPermission remoteIndicesPermission;
+        try {
+            remoteIndicesPermission = role.remoteIndices().forCluster(remoteClusterAlias);
+        } catch (UnsupportedOperationException e) {
+            // TODO we will need to implement this to support API keys with assigned role descriptors
+            final String message = "Cannot retrieve remote access role descriptors for API keys with assigned role descriptors.";
+            assert false : message;
+            listener.onFailure(new IllegalArgumentException(message, e));
```

Comment:
Do we need assert here? It is not a programming error but rather user choice (of using ApiKey authentication) if code reaches here. This is similar to [throwing exception](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java#L309) when creating API key with remote indices privileges and we did not assert there. So I'd suggest we remove the `assert` and optionally add a test for the failure scenario.

Reply:
Good point that we should be consistent about how we disallow API keys -- since we will need to add support for these soon, I don't feel strongly either way. I'll remove the assertion.